driver-core/src/main/com/mongodb/operation/OperationHelper.java

```diff
@@ -444,6 +526,76 @@ static void withConnection(final AsyncReadBinding binding, final AsyncCallableWi
+        return withConnection(binding, new CallableWithConnection<T>() {
+            @Override
+            public T call(final Connection connection) {
+                try {
+                    return callable.call(connection, txnNumber);
+                } catch (MongoException e) {
+                    if (!retryable) {
+                        throw e;
+                    } else {
+                        final MongoException originalException = e;
+                        return withConnection(binding, new CallableWithConnection<T>() {
+                            @Override
+                            public T call(final Connection connection) {
+                                if (!serverIsAtLeastVersionThreeDotSix(connection.getDescription())) {
+                                    throw originalException;
+                                }
+                                return callable.call(connection, txnNumber);
+                            }
+                        });
+                    }
+                }
+            }
+        });
+    }
+
+    static <T> void retryableWithConnection(final AsyncWriteBinding binding, final boolean retryable,
+                                            final SingleResultCallback<T> callback,
+                                            final AsyncCallableWithConnectionAndTransactionNumber<T> callable) {
+        final SingleResultCallback<T> errHandlingCallback = errorHandlingCallback(callback, LOGGER);
+        final Long txnNumber = retryable ? binding.getSessionContext().advanceTransactionNumber() : null;
+        withConnection(binding, new AsyncCallableWithConnection() {
+            @Override
+            public void call(final AsyncConnection connection, final Throwable t) {
+                if (t != null) {
+                    errHandlingCallback.onResult(null, t);
+                } else {
+                    callable.call(connection, txnNumber, new SingleResultCallback<T>() {
+                        @Override
+                        public void onResult(final T result, final Throwable originalError) {
+                            if (originalError != null) {
+                                if (!retryable || !(originalError instanceof MongoException)) {
+                                    releasingCallback(errHandlingCallback, connection).onResult(null, originalError);
+                                } else {
+                                    connection.release();
+                                    withConnection(binding, new AsyncCallableWithConnection() {
+                                        @Override
+                                        public void call(final AsyncConnection newConnection, final Throwable t) {
+                                            if (t != null) {
+                                                errHandlingCallback.onResult(null, originalError);
```

Comment:
I think this is inconsistent with the corresponding synchronous code.  The synchronous code appears to throw any exception from server selection/connection pool on the retry attempt, whereas here the original error is "thrown" asynchronously.  The spec's pseudocode implies that the synchronous code is correct, since there is no try catch around the selectServer call.

Reply:
I'm of the opinion though that this behavior is actually better.